### PR TITLE
L1 sync

### LINKED ISSUE #73: Ref Impl: Service to retrieve blocks with full receipt, to get logs from a layer1 node
Transferred from ENG-1739

### PR DESCRIPTION
**Pull request by protolambda**
Changes:
- Write a "L1 tracker"
  - Resolves reorgs
  - Used to schedule the next block to process on top of any specific local view (to support multiple engines, and be more robust)
  - Watches for head changes to update tracker
  - Also polls to ensure nothing is missed if subscription is temporarily shaky
- Write a "L1 downloader"
  - Fetches blocks concurrently
  - Fetches receipts concurrently, with parallel retrieval of receipts, split between workers
  - de-duplicate requests (TODO: needs pruning of old requests)
- Organize L1 nodes as "sources"
  - Easier to mock a source interface than a full L1 node
  - L1 sources can be combined (TODO: head changes of only the first are respected, but downloads etc. are split). Maybe we'll optimize further later,  but at least the design can handle multiple sources
- Organize L2 execution engines as "engines"
  - Keep all engines synced, even if they start at different places
  - Useful to have L2 redundancy as sequencer etc. (note that beacon nodes also support attaching multiple execution engines)
- More node organization: engines, tracker, downloader etc. are all connected (subscriptions etc. are set up and torn down on close)

Status: discuss, some review, then merge into `staging`. Then driver PR next, and integration tests?

Fix #73

### CODE REVIEW

**Comment by protolambda** [comment]
Merging into `staging`, Hopefully JG can review after holidays

File: opnode/node/node.go [hunk 1]
```diff
@@ -2,85 +2,175 @@ package node
...
+	"github.com/ethereum/go-ethereum/ethclient"
+
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
 type OpNodeCmd struct {
-	L1NodeAddr   string `ask:"--l1" help:"Address of L1 User JSON-RPC endpoint to use (eth namespace required)"`
-	L2EngineAddr string `ask:"--l2" help:"Address of L2 Engine JSON-RPC endpoint to use (engine and eth namespace required)"`
+	L1NodeAddrs   []string `ask:"--l1" help:"Addresses of L1 User JSON-RPC endpoints to use (eth namespace required)"`
+	L2EngineAddrs []string `ask:"--l2" help:"Addresses of L2 Engine JSON-RPC endpoints to use (engine and eth namespace required)"`
```
Comment (norswap) [inline opnode/node/node.go:21]: Would be great to document the format of JSON-RPC endpoints (what are the engine and eth namespaces?) and giving an example. Can be in markdown doc and not in code.

File: opnode/node/node.go [hunk 2]
```diff
@@ -2,85 +2,175 @@ package node
...
+
 	c.close = make(chan chan error)
 
 	go c.RunNode()
 
 	return nil
 }
 
-// background process
 func (c *OpNodeCmd) RunNode() {
```
Comment (norswap) [inline opnode/node/node.go:111]: - What is the distinction between `Run`  and `RunNode`?
- Must `RunNode` be public?

**Review by norswap — COMMENTED** [review]
Haven't reviewed downloader, tracker, etc, but posting what I have.

**Review by norswap — COMMENTED** [review]
Extra comment : In the documentation for `ForkchoiceState`, the comments are not useful if they don't say what "safe" and "finalized" mean.

File: opnode/l2/engine.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (c *Engine) ProcessL1(dl l1.Downloader, finalized common.Hash, id l1.BlockID) {
+	if id == c.L1Head {
+		// no-op, already processed it
+		return
+	}
+	// check if it's in the far distance.
+	// Header sync will need to move closer before we start fetching/caching full blocks.
+	if id.Number > c.L1Head.Number+5 {
```
[the remark this answers is not in the document]
Comment (protolambda) [reply opnode/l2/engine.go:32]: This code is removed in l2-tracking PR with the fixed sync code #105

**Pull request merged**